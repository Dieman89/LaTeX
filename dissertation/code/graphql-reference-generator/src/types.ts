import {
  GraphQLNamedType,
  GraphQLArgument,
  GraphQLField,
  GraphQLDirective,
} from "graphql";
import {
  DIRECTIVES,
  ENUMS,
  INPUTS,
  INTERFACES,
  MUTATIONS,
  OBJECTS,
  QUERIES,
  SCALARS,
  SUBSCRIPTIONS,
  UNIONS,
} from "./constants";

// can be any type that is defined in the schema or a component of those types (eg. a field or an argument)
export type GraphQLItem = {
  name: string; // could be the name of the type or an alias if it's a field or argument
  description: string; // the description of the item from the schema file
  type: string | null; // name of the item type if it is not a type itself
  printed: string | null; // a string showing the item in GraphQL schema format
  arguments: Array<GraphQLItem>; // any arguments that the item takes
  fields: Array<GraphQLItem>; // any fields on the item
  path: string | null; // the path to reach the type in the GraphQLItemData
  isRootField?: boolean; // `true` if the item is a field of 'Query', 'Mutation' or 'Subscription' types
  directives: Array<GraphQLDirectiveItem>;
};

export type GraphQLDirectiveItem = Pick<GraphQLItem, "path" | "name">;

export type GraphQLRawItem =
  | GraphQLNamedType
  | GraphQLArgument
  | GraphQLDirective
  | GraphQLField<any, any, any>;

export type ReferenceKeys =
  | typeof QUERIES
  | typeof MUTATIONS
  | typeof SUBSCRIPTIONS
  | typeof OBJECTS
  | typeof ENUMS
  | typeof SCALARS
  | typeof INTERFACES
  | typeof UNIONS
  | typeof INPUTS
  | typeof DIRECTIVES;

export type GraphQLReferenceGroup = {
  [key in ReferenceKeys]: Array<GraphQLItem>;
};
