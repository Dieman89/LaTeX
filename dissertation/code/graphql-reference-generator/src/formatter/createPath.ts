import {
  isInterfaceType,
  isObjectType,
  isEnumType,
  isScalarType,
  isUnionType,
  isInputType,
  isListType,
  isNonNullType,
  GraphQLType,
  DirectiveNode,
} from "graphql";
import {
  DIRECTIVES,
  ENUMS,
  INPUTS,
  INTERFACES,
  OBJECTS,
  SCALARS,
  UNIONS,
} from "../constants";
import { toKebabCase } from "../utils/toKebabCase";

export const createPathToItem = (type: GraphQLType | DirectiveNode): string => {
  if (isListType(type) || isNonNullType(type)) {
    return createPathToItem(type.ofType);
  }
  if (isScalarType(type)) {
    return `/${SCALARS}/${toKebabCase(type.name)}`;
  }
  if (isObjectType(type)) {
    return `/${OBJECTS}/${toKebabCase(type.name)}`;
  }
  if (isEnumType(type)) {
    return `/${ENUMS}/${toKebabCase(type.name)}`;
  }
  if (isInterfaceType(type)) {
    return `/${INTERFACES}/${toKebabCase(type.name)}`;
  }
  if (isUnionType(type)) {
    return `/${UNIONS}/${toKebabCase(type.name)}`;
  }
  if (isInputType(type)) {
    return `/${INPUTS}/${toKebabCase(type.name)}`;
  }
  return `/${DIRECTIVES}/${toKebabCase(type.name.value)}`;
};
