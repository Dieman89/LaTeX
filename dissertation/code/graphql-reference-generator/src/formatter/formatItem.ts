import { GraphQLDirectiveItem, GraphQLItem, GraphQLRawItem } from "../types";
import {
  GraphQLArgument,
  GraphQLField,
  GraphQLObjectType,
  GraphQLInterfaceType,
  GraphQLInputObjectType,
  print,
  ASTNode,
  DirectiveNode,
  GraphQLDirective,
} from "graphql";
import { createPathToItem } from "./createPath";

export const formatItem = (graphqlType: GraphQLRawItem): GraphQLItem => {
  return {
    name: graphqlType.name,
    description: graphqlType.description || "",
    type: hasType(graphqlType) ? graphqlType.type.toString() : null,
    printed: printASTNode(graphqlType.astNode as ASTNode),
    arguments: hasArgs(graphqlType)
      ? Object.values(graphqlType.args).map(formatItem)
      : [],
    fields: hasFields(graphqlType)
      ? Object.values(graphqlType.getFields()).map(formatItem)
      : [],
    path: hasType(graphqlType)
      ? createPathToItem(graphqlType.type).toLowerCase()
      : null,
    directives: hasDirectives(graphqlType)
      ? createDirectives(graphqlType.astNode.directives)
      : [],
  };
};

export function printASTNode(astNode: ASTNode) {
  const pattern = new RegExp(/"""([\s\S]*?)"""\n/, "g");
  return astNode
    ? print(astNode).replace(pattern, "").trim().replace(/^ +/gm, "")
    : null;
}

export function createDirectives(
  nodes: Array<DirectiveNode>
): Array<GraphQLDirectiveItem> {
  return nodes.map((item: DirectiveNode) => {
    return {
      name: item.name.value,
      path: createPathToItem(item),
    };
  });
}

function hasType(
  type: any
): type is GraphQLArgument | GraphQLField<any, any, any> {
  return !!type.type;
}

function hasArgs(
  type: any
): type is GraphQLField<any, any, any> | GraphQLDirective {
  return !!type.args;
}

function hasFields(
  type: any
): type is GraphQLObjectType | GraphQLInterfaceType | GraphQLInputObjectType {
  return !!type.getFields;
}

function hasDirectives(
  type: any
): type is { astNode: { directives: DirectiveNode[] } } {
  return !!(type.astNode?.directives?.length > 0);
}
