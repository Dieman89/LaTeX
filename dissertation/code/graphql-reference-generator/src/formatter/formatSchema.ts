import {
  GraphQLDirective,
  GraphQLNamedType,
  GraphQLObjectType,
  GraphQLSchema,
  isEnumType,
  isInputType,
  isInterfaceType,
  isObjectType,
  isScalarType,
  isUnionType,
} from "graphql";
import {
  ENUMS,
  INPUTS,
  INTERFACES,
  MUTATIONS,
  OBJECTS,
  QUERIES,
  SCALARS,
  SUBSCRIPTIONS,
  UNIONS,
  DIRECTIVES,
} from "../constants";
import { formatItem } from "./formatItem";
import { GraphQLReferenceGroup } from "../types";

export function formatSchema(schema: GraphQLSchema) {
  return {
    ...Object.values(schema.getTypeMap())
      .filter((type) => !type.name.startsWith("__"))
      .reduce(typeReducer, initialItemData),
    directives: schema.getDirectives().map(formatItem),
  };
}

const initialItemData: GraphQLReferenceGroup = {
  [QUERIES]: [],
  [MUTATIONS]: [],
  [SUBSCRIPTIONS]: [],
  [OBJECTS]: [],
  [ENUMS]: [],
  [SCALARS]: [],
  [INTERFACES]: [],
  [UNIONS]: [],
  [INPUTS]: [],
  [DIRECTIVES]: [],
};

function typeReducer(
  acc: GraphQLReferenceGroup,
  type: GraphQLNamedType | GraphQLDirective
): GraphQLReferenceGroup {
  if (isObjectType(type)) {
    switch (type.name) {
      case "Query":
        return {
          ...acc,
          [QUERIES]: createRootObjectArray(type),
        };

      case "Mutation":
        return {
          ...acc,
          [MUTATIONS]: createRootObjectArray(type),
        };

      case "Subscription":
        return {
          ...acc,
          [SUBSCRIPTIONS]: createRootObjectArray(type),
        };

      default:
        return {
          ...acc,
          [OBJECTS]: [...acc[OBJECTS], formatItem(type)],
        };
    }
  }

  if (isEnumType(type)) {
    return {
      ...acc,
      [ENUMS]: [...acc[ENUMS], formatItem(type)],
    };
  }

  if (isScalarType(type)) {
    return {
      ...acc,
      [SCALARS]: [...acc[SCALARS], formatItem(type)],
    };
  }

  if (isInterfaceType(type)) {
    return {
      ...acc,
      [INTERFACES]: [...acc[INTERFACES], formatItem(type)],
    };
  }

  if (isUnionType(type)) {
    return {
      ...acc,
      [UNIONS]: [...acc[UNIONS], formatItem(type)],
    };
  }

  if (isInputType(type)) {
    return {
      ...acc,
      [INPUTS]: [...acc[INPUTS], formatItem(type)],
    };
  }

  return acc;
}

function createRootObjectArray(type: GraphQLObjectType) {
  return Object.values(type.getFields())
    .map(formatItem)
    .map((formattedType) => ({
      ...formattedType,
      isRootField: true,
    }));
}
