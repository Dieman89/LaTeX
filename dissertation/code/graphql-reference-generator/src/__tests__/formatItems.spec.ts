import { formatItem, printASTNode } from "../formatter/formatItem";
import {
  ASTNode,
  buildSchema,
  GraphQLDirective,
  GraphQLObjectType,
  isEnumType,
  isInputType,
  isInterfaceType,
  isObjectType,
  isScalarType,
  isUnionType,
} from "graphql";
import { GraphQLRawItem } from "../types";

describe("formatType Function tests", () => {
  it("should give a nice well formatted query object field", () => {
    const schema = buildSchema(`
      type Query {
        """This is a test description"""
        testQueryRoot(testArgument: String): String
      }
    `);
    const queryFields = (
      Object.values(schema.getTypeMap()).find(
        (type: { name: string }) => type.name === "Query"
      )! as GraphQLObjectType
    ).getFields();
    const queryField = Object.values(queryFields)[0] as GraphQLRawItem;

    const output = formatItem(queryField);

    expect(output).toMatchSnapshot();
  });

  it("should give a nice well formatted mutation object field", () => {
    const schema = buildSchema(`
      type Mutation {
        """Test mutation description"""
        testMutation(testMutationArgument: TestInput): String
      }

      input TestInput {
        id: String
      }
    `);
    const queryItems = (
      Object.values(schema.getTypeMap()).find(
        (type: { name: string }) => type.name === "Mutation"
      )! as GraphQLObjectType
    ).getFields();
    const queryField = Object.values(queryItems)[0] as GraphQLRawItem;
    console.log(queryField);

    const output = formatItem(queryField);
    console.log(output);
    expect(output).toMatchSnapshot();
  });

  it("should give a nice well formatted subscription object field", () => {
    const schema = buildSchema(`
      type Subscription {
        """Test subscription description"""
        testSubscription(testSubscriptionArgument: TestInput): String
      }

      input TestInput {
        id: String
      }
    `);
    const queryItems = (
      Object.values(schema.getTypeMap()).find(
        (type: { name: string }) => type.name === "Subscription"
      )! as GraphQLObjectType
    ).getFields();
    const queryField = Object.values(queryItems)[0]! as GraphQLRawItem;

    const output = formatItem(queryField);

    expect(output).toMatchSnapshot();
  });

  it("should provide the correct format from a GraphQL object type", () => {
    const schema = buildSchema(`
      type TestObject {
        aString: String
        aNonNullString: String!
        aNumber: Int
        aNestedType: TestNestedType!
      }

      type TestNestedType {
        anotherString: String
      }
    `);
    const objectToTest = Object.values(schema.getTypeMap()).find(
      (type) => isObjectType(type) && type.name === "TestObject"
    )!;

    const output = formatItem(objectToTest);

    expect(output).toMatchSnapshot();
  });

  it("should give a nice well formatted enum object", () => {
    const schema = buildSchema(`enum TestEnum { TEST1 TEST2 TEST3 }`);
    const enumToTest = Object.values(schema.getTypeMap()).find(
      (type: { name: string }) => isEnumType(type) && type.name === "TestEnum"
    )!;

    const output = formatItem(enumToTest);

    expect(output).toMatchSnapshot();
  });

  it("should provide the correct format from a GraphQL scalar type", () => {
    const schema = buildSchema(`
      scalar TestScalar
    `);
    const scalarToTest = Object.values(schema.getTypeMap()).find(
      (type) => isScalarType(type) && type.name === "TestScalar"
    )!;

    const output = formatItem(scalarToTest);

    expect(output).toMatchSnapshot();
  });

  it("should provide the correct format from a GraphQL union type", () => {
    const schema = buildSchema(`
      union TestUnion = String | Int
    `);
    const unionToTest = Object.values(schema.getTypeMap()).find(
      (type) => isUnionType(type) && type.name === "TestUnion"
    )!;

    const output = formatItem(unionToTest);

    expect(output).toMatchSnapshot();
  });

  it("should give a nice well formatted interface object", () => {
    const schema = buildSchema(`
      interface TestInterface { test: String }
    `);
    const interfaceToTest = Object.values(schema.getTypeMap()).find(
      (type: { name: string }) =>
        isInterfaceType(type) && type.name === "TestInterface"
    )!;

    const output = formatItem(interfaceToTest);

    expect(output).toMatchSnapshot();
  });

  it("should give a nice well formatted input object", () => {
    const schema = buildSchema(`
      input TestInput { test: String }
    `);
    const inputToTest = Object.values(schema.getTypeMap()).find(
      (type: { name: string }) => isInputType(type) && type.name === "TestInput"
    )!;

    const output = formatItem(inputToTest);

    expect(output).toMatchSnapshot();
  });

  it("should give a nice well formatted directive object", () => {
    const schema = buildSchema(`
      directive @testDirective(
        testArg: String = "Test argument"
      ) on FIELD_DEFINITION
    `);
    const directiveToTest = Object.values(schema.getDirectives()).find(
      (type: GraphQLDirective) => type.name === "testDirective"
    )!;

    const output = formatItem(directiveToTest);

    expect(output).toMatchSnapshot();
  });

  it("should remove all the comments from a printed node", () => {
    const schema = buildSchema(`
      """Interface Comment"""
      interface TestInterface {
        """Interface Field Comment"""
        test: String
      }
    `);
    const interfaceObject = Object.values(schema.getTypeMap()).find(
      isInterfaceType
    )!;

    const output = printASTNode(interfaceObject.astNode! as ASTNode);

    expect(output).not.toMatch(/"""([\s\S]*?)"""\n/);
  });

  it("should provide the type for arguments or fields", () => {
    const schema = buildSchema(`
      type Query {
        testField(testArgument: TestObject): TestObject
      }

      type TestObject {
        testField: String
      }
    `);
    const typeMap = Object.values(schema.getTypeMap());
    const queryFields = (
      typeMap.find(
        (type: { name: string }) => type.name === "Query"
      )! as GraphQLObjectType
    ).getFields();
    const queryFieldToTest = Object.values(queryFields)[0] as GraphQLRawItem;

    const output = formatItem(queryFieldToTest);

    expect(output.type).toEqual("TestObject");
  });

  it("should not provide the type named types", () => {
    const schema = buildSchema(`
      type TestObject {
        testField: String
      }
    `);
    const typeMap = Object.values(schema.getTypeMap());
    const objectToTest = typeMap.find(
      (type) => isObjectType(type) && type.name === "TestObject"
    )!;

    const output = formatItem(objectToTest);

    expect(output.type).toBeNull();
  });
});
