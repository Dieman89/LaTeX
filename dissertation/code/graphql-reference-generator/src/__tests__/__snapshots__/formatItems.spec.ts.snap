// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatType Function tests should give a nice well formatted directive object 1`] = `
Object {
  "arguments": Array [
    Object {
      "arguments": Array [],
      "description": "",
      "directives": Array [],
      "fields": Array [],
      "name": "testArg",
      "path": "/scalars/string",
      "printed": "testArg: String = \\"Test argument\\"",
      "type": "String",
    },
  ],
  "description": "",
  "directives": Array [],
  "fields": Array [],
  "name": "testDirective",
  "path": null,
  "printed": "directive @testDirective(testArg: String = \\"Test argument\\") on FIELD_DEFINITION",
  "type": null,
}
`;

exports[`formatType Function tests should give a nice well formatted enum object 1`] = `
Object {
  "arguments": Array [],
  "description": "",
  "directives": Array [],
  "fields": Array [],
  "name": "TestEnum",
  "path": null,
  "printed": "enum TestEnum {
TEST1
TEST2
TEST3
}",
  "type": null,
}
`;

exports[`formatType Function tests should give a nice well formatted input object 1`] = `
Object {
  "arguments": Array [],
  "description": "",
  "directives": Array [],
  "fields": Array [
    Object {
      "arguments": Array [],
      "description": "",
      "directives": Array [],
      "fields": Array [],
      "name": "test",
      "path": "/scalars/string",
      "printed": "test: String",
      "type": "String",
    },
  ],
  "name": "TestInput",
  "path": null,
  "printed": "input TestInput {
test: String
}",
  "type": null,
}
`;

exports[`formatType Function tests should give a nice well formatted interface object 1`] = `
Object {
  "arguments": Array [],
  "description": "",
  "directives": Array [],
  "fields": Array [
    Object {
      "arguments": Array [],
      "description": "",
      "directives": Array [],
      "fields": Array [],
      "name": "test",
      "path": "/scalars/string",
      "printed": "test: String",
      "type": "String",
    },
  ],
  "name": "TestInterface",
  "path": null,
  "printed": "interface TestInterface {
test: String
}",
  "type": null,
}
`;

exports[`formatType Function tests should give a nice well formatted mutation object field 1`] = `
Object {
  "arguments": Array [
    Object {
      "arguments": Array [],
      "description": "",
      "directives": Array [],
      "fields": Array [],
      "name": "testMutationArgument",
      "path": "/inputs/test-input",
      "printed": "testMutationArgument: TestInput",
      "type": "TestInput",
    },
  ],
  "description": "Test mutation description",
  "directives": Array [],
  "fields": Array [],
  "name": "testMutation",
  "path": "/scalars/string",
  "printed": "testMutation(testMutationArgument: TestInput): String",
  "type": "String",
}
`;

exports[`formatType Function tests should give a nice well formatted query object field 1`] = `
Object {
  "arguments": Array [
    Object {
      "arguments": Array [],
      "description": "",
      "directives": Array [],
      "fields": Array [],
      "name": "testArgument",
      "path": "/scalars/string",
      "printed": "testArgument: String",
      "type": "String",
    },
  ],
  "description": "This is a test description",
  "directives": Array [],
  "fields": Array [],
  "name": "testQueryRoot",
  "path": "/scalars/string",
  "printed": "testQueryRoot(testArgument: String): String",
  "type": "String",
}
`;

exports[`formatType Function tests should give a nice well formatted subscription object field 1`] = `
Object {
  "arguments": Array [
    Object {
      "arguments": Array [],
      "description": "",
      "directives": Array [],
      "fields": Array [],
      "name": "testSubscriptionArgument",
      "path": "/inputs/test-input",
      "printed": "testSubscriptionArgument: TestInput",
      "type": "TestInput",
    },
  ],
  "description": "Test subscription description",
  "directives": Array [],
  "fields": Array [],
  "name": "testSubscription",
  "path": "/scalars/string",
  "printed": "testSubscription(testSubscriptionArgument: TestInput): String",
  "type": "String",
}
`;

exports[`formatType Function tests should provide the correct format from a GraphQL object type 1`] = `
Object {
  "arguments": Array [],
  "description": "",
  "directives": Array [],
  "fields": Array [
    Object {
      "arguments": Array [],
      "description": "",
      "directives": Array [],
      "fields": Array [],
      "name": "aString",
      "path": "/scalars/string",
      "printed": "aString: String",
      "type": "String",
    },
    Object {
      "arguments": Array [],
      "description": "",
      "directives": Array [],
      "fields": Array [],
      "name": "aNonNullString",
      "path": "/scalars/string",
      "printed": "aNonNullString: String!",
      "type": "String!",
    },
    Object {
      "arguments": Array [],
      "description": "",
      "directives": Array [],
      "fields": Array [],
      "name": "aNumber",
      "path": "/scalars/int",
      "printed": "aNumber: Int",
      "type": "Int",
    },
    Object {
      "arguments": Array [],
      "description": "",
      "directives": Array [],
      "fields": Array [],
      "name": "aNestedType",
      "path": "/objects/test-nested-type",
      "printed": "aNestedType: TestNestedType!",
      "type": "TestNestedType!",
    },
  ],
  "name": "TestObject",
  "path": null,
  "printed": "type TestObject {
aString: String
aNonNullString: String!
aNumber: Int
aNestedType: TestNestedType!
}",
  "type": null,
}
`;

exports[`formatType Function tests should provide the correct format from a GraphQL scalar type 1`] = `
Object {
  "arguments": Array [],
  "description": "",
  "directives": Array [],
  "fields": Array [],
  "name": "TestScalar",
  "path": null,
  "printed": "scalar TestScalar",
  "type": null,
}
`;

exports[`formatType Function tests should provide the correct format from a GraphQL union type 1`] = `
Object {
  "arguments": Array [],
  "description": "",
  "directives": Array [],
  "fields": Array [],
  "name": "TestUnion",
  "path": null,
  "printed": "union TestUnion = String | Int",
  "type": null,
}
`;
